### PR TITLE
Fix token validation

### PR DESCRIPTION
**Pull request by HugoPouliquen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,6 +53,40 @@ hotp.at(1401) # => "316439"
...
+  config.opt_mem_cache = :redis
+  config.opt_mem_cache_conf = {
+    url: 'redis://localhost:6379',
+    db: 10,
+    expiration_time: 300
+  }
+end
+```
+
+#### Use
```
Comment (mdouchement) [inline README.md]: - Usage

File: README.md [hunk 2]
```diff
@@ -53,6 +53,40 @@ hotp.at(1401) # => "316439"
...
+
+```bash
+totp = ROTP::TOTP.new("base32secret3232")
+totp.now # => "492039"
+
+totp.verify("492039", uid: 1) # => true
+# less than 30 secs
+totp.verify("492039", uid: 1) # => fals
```
Comment (mdouchement) [inline README.md]: - `false`

File: lib/rotp/opt_mem_caches/redis.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      attr_reader :redis, :expiration_time
+
+      def initialize
+        @redis = ::Redis.new(
+          url: ROTP.config.opt_mem_cache_conf[:url],
+          db: ROTP.config.opt_mem_cache_conf[:db]
+        )
+        RedisClassy.redis = @redis
+        @expiration_time = ROTP.config.opt_mem_cache_conf[:expiration_time]
+        self
```
Comment (mdouchement) [inline lib/rotp/opt_mem_caches/redis.rb]: - Useless `self` keyword

File: lib/rotp/otp.rb [hunk 4]
```diff
@@ -41,11 +42,13 @@ def generate_otp(input, padded=true)
...
 
-    def verify(input, generated)
+    def verify(input, generated, uid:)
       unless input.is_a?(String) && generated.is_a?(String)
         raise ArgumentError, "ROTP only verifies strings - See: https://github.com/mdp/rotp/issues/32"
       end
-      time_constant_compare(input, generated)
+      
+      time_constant_compare(input, generated) &&
+        opt_mem_cache.unused?(uid, input)
```
Comment (mdouchement) [inline lib/rotp/otp.rb]: - This line raise a `NoMethodError` when `ROTP.config.opt_mem_cache` is `nil`

File: spec/lib/rotp/opt_mem_caches/redis_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        after do
+          ROTP.config.opt_mem_cache_conf = @opt_mem_cache_conf_backup
+        end
+
+        it 'is unused' do
+          expect(subject.unused?(uid, otp)).to eq(true)
+        end
+
+        it 'is expired' do
```
Comment (mdouchement) [inline spec/lib/rotp/opt_mem_caches/redis_spec.rb]: - adds a TTL on otp

File: spec/lib/rotp/totp_spec.rb [hunk 6]
```diff
@@ -80,6 +80,41 @@
...
+            expect(totp.verify(input, uid: uid)).to be_falsey
+          end
+        end
+      end
+
+
+      context 'when time constant compare is invalid' do
+        let(:input) { '123456' }
+        let(:uid)   { 1 }
+        it 'is not called' do
```
Comment (mdouchement) [inline spec/lib/rotp/totp_spec.rb]: - add a new line ;p

File: spec/lib/rotp/totp_spec.rb [hunk 7]
```diff
@@ -80,6 +80,41 @@
...
+          it 'is false' do
+            expect(totp.verify(input, uid: uid)).to be_falsey
+          end
+        end
+      end
+
+
+      context 'when time constant compare is invalid' do
+        let(:input) { '123456' }
+        let(:uid)   { 1 }
```
Comment (mdouchement) [inline spec/lib/rotp/totp_spec.rb]: - add a new line ;p

File: spec/lib/rotp/totp_spec.rb [hunk 8]
```diff
@@ -80,6 +80,41 @@
           expect(totp.verify('102705')).to be_falsey
         end
       end
+
+      context 'when time constant compare is valid' do
+        let(:input) { '102705' }
+        let(:uid)   { 1 }
+        let(:now)   { Time.at 1_297_553_958 }
+
+        context 'when unused? is true ' do
```
Comment (mdouchement) [inline spec/lib/rotp/totp_spec.rb]: - when otp is not used

File: spec/lib/rotp/totp_spec.rb [hunk 9]
```diff
@@ -80,6 +80,41 @@
...
+          before do
+            allow(totp.opt_mem_cache).to receive(:unused?).with(uid, input).and_return(true)
+          end
+
+          it 'is true' do
+            expect(totp.verify(input, uid: uid)).to be_truthy
+          end
+        end
+
+        context 'when unused? is false ' do
```
Comment (mdouchement) [inline spec/lib/rotp/totp_spec.rb]: - when otp is already used

File: spec/lib/rotp/opt_mem_caches/redis_spec.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    describe '#unused?' do
+      let(:otp) { '354256' }
+      let(:uid) { 1 }
+      context 'when there is no uid' do
+        it 'returns true' do
+          expect(subject.unused?(nil, nil)).to eq(true)
+        end
+      end
+
+      context 'when otp is not stored in the opt mem cache' do
```
Comment (HugoPouliquen) [inline spec/lib/rotp/opt_mem_caches/redis_spec.rb]: not present

**Pull request closed without merging**